drivers/pci/pci.c

```diff
@@ -206,6 +207,30 @@ void LpcSetSerialState(int enable)
 	LpcWriteRegister(0x30, enable ? 0x01 : 0x00);
 }
 
+int serial_putchar(int ch)
+{
+	/* Wait for THRE (bit 5) to be high */
+	while ((IoInputByte(SERIAL_PORT + SERIAL_LSR) & (1 << 5)) == 0);
+	IoOutputByte(SERIAL_PORT + SERIAL_THR, ch);
+	return ch;
+}
+
+void bprintf(const char *fmt, ...)
+{
+	va_list args;
+	char buf[256] = {0};
+	int i;
+
+	va_start(args, fmt);
+	vsprintf(&buf, fmt, args);
```

Comment:
Any formatted string of length greater than 255 will overflow buffer. Unfortunately, it looks like we don't have a 'vsnprintf' implementation that would allow us to specify the size of the buffer. It's probably ok for now, but something to watch out for and which should likely be fixed in the future.